6901 support tiled object type map property

JohnLavender474 opened the pull request:
This PR is in response to issue [6901 Support Tiled "object" type MapProperty](https://github.com/libgdx/libgdx/issues/6901). The request is that the "object" property type in Tiled be supported in LibGDX. Here, my solution is to use the MapProperties of the object being pointed to as the value of the "object" type property. 

In the [BaseTmxMapLoader](https://github.com/JohnLavender474/LibGDX--Support-Tiled-object-type-MapProperty-6901/blob/6901_Support-Tiled-object-type-MapProperty/gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java) class, the properties of objects are loaded as normal except for a few differences. 

The first difference is that mappings are added to an ObjectMap instance where the key is the id of the object and the value is the MapProperties of the object. 

The second difference is that when the type of a property is "object", then a Runnable instance is created where in the run method the MapProperties of the object being pointed to is added as a property. The Runnable is added to a queue, and the Runnables are not run until the end of the loadTiledMap method. Doing it this way ensures that all MapProperties are loaded before attempting to fetch them.

The [TiledMapObjectPropertyTest](https://github.com/JohnLavender474/LibGDX--Support-Tiled-object-type-MapProperty-6901/blob/6901_Support-Tiled-object-type-MapProperty/tests/gdx-tests/src/com/badlogic/gdx/tests/TiledMapObjectPropertyTest.java) class tests that "object" properties are successfully loaded. The test includes ensuring that a circular graph does not lead to infinite recursion. Feel free to open [the tmx file](https://github.com/JohnLavender474/LibGDX--Support-Tiled-object-type-MapProperty-6901/blob/6901_Support-Tiled-object-type-MapProperty/tests/gdx-tests-android/assets/data/maps/tiled-objects/test-object-properties.tmx) to ensure proper testing.

JohnLavender474 commented:
@SimonIT Thanks for looking over my PR. How many reviews should it receive before it's greenlit to merge into the master branch?

SimonIT commented:
Our rule of thumb is two reviews per PR. Good luck finding someone who reviews the PR. We are searching for some time now for someone who reviews tiled PRs

tommyettinger commented on tests/gdx-tests/src/com/badlogic/gdx/tests/TiledMapObjectPropertyTest.java:
```diff
@@ -0,0 +1,190 @@
+		shapeRenderer.dispose();
+	}
+
+	private void test (MapProperties props, int idToObjProp, ObjectMap<Integer, MapProperties> idPropMap) {
```
This would probably also need changing if you do switch the other `ObjectMap<Integer, MapProperties>` to use an IntMap.

tommyettinger reviewed (changes requested):
Looks pretty good; I have some minor suggestions, but I don't think anything should be hard to implement. I'm not really sure what our variable name code style rules are, other than I don't see underscore prefixes in many places in libGDX. I think switching to IntMap is only really something I will push for in BaseTmxMapLoader.java ; I think it's a good idea in the test but isn't necessary.

Yay, Queue! Another poorly-named class with a nice implementation! I just wanted to comment that I hadn't seen it used much, and I'm glad it is getting used now.

tommyettinger commented on tests/gdx-tests/src/com/badlogic/gdx/tests/TiledMapObjectPropertyTest.java:
```diff
@@ -0,0 +1,190 @@
+
+			MapObjects objects = map.getLayers().get("Objects").getObjects();
+
+			ObjectMap<Integer, MapProperties> idPropMap = new ObjectMap<>();
```
I forgot to post that this is the other usage of ObjectMap with Integer keys, and could be an IntMap.

JohnLavender474 commented on gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java:
```diff
@@ -430,23 +441,59 @@ protected void loadObject (TiledMap map, MapObjects objects, Element element, fl
+						 * Value should be the id of the object being pointed to. Nevertheless, the try-catch block safeguards in case
+						 * faulty data in inputted.
+						 */
+						final int _value = Integer.parseInt(value);
```
Good suggestion. I changed the name to id. And yeah, I'm glad to make use of the LibGDX collections! I recently wrote a simple game engine in Kotlin built on LibGDX. One thing on my ToDo list is to go back through that project and swap out all the Java collections with LibGDX ones. The optimization is definitely worth the effort!

JohnLavender474 commented on gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java:
```diff
@@ -64,6 +65,10 @@ public static class Parameters extends AssetLoaderParameters<TiledMap> {
 
 	protected TiledMap map;
+	/** MapProperty instances are mapped to the id of the respective tiled object */
+	protected ObjectMap<Integer, MapProperties> idToPropsMap;
```
Changed the usages of ObjectMap to IntMap. Again, good suggestion! Never hurts to get things as optimized as possible.

JohnLavender474 commented:
@tommyettinger Didn't mean to dismiss your review. I wanted to make a few more changes. I renamed the map in BaseTmxMapLoader to "objectIdToPropsMap" to make it clear that the map ONLY contains [id, MapProperties] entries for objects (and not tiles, animated tiles, or the Tiled map). Also, I cleaned up the logs in the TiledMapObjectPropertyTest file.

tommyettinger commented on tests/gdx-tests/src/com/badlogic/gdx/tests/TiledMapObjectPropertyTest.java:
```diff
@@ -135,9 +133,11 @@ public void create () {
 					}
 
-					System.out.println("\t" + key + " = " + value);
+					builder.append("\t").append(key).append(" = ").append(value).append("\n");
```
This is definitely cleaner than mixing log() and println() calls.

It doesn't matter here at all, but when anyone else reviews this, I'm wondering when the StringBuilder class in libGDX is considered better than StringBuilder in the JDK. The JDK one has a number of intrinsics, if I remember it right, and also has better support from IDE tooling. The libGDX one supports equals() and hashCode().

Also, who keeps coming up with duplicate names? Array, Queue, and StringBuilder are all long-existing names from the JDK. At least StringBuilder actually correctly describes what the class is. `Array<T>` is much closer to an `ArrayList<T>` than a `T[]`, while `Queue` is just as much a stack as it is a queue, and can almost implement `Deque` from the JDK (for double-ended queues). Even AssetManager has some confusing code in AndroidFileHandle that uses Android's AssetManager class and not the libGDX one.

NathanSweet replied:
StringBuilder collects chars and exposes the array, it has hascode, equals, and a couple utility methods. IIRC the JRE class stores bytes and doesn't expose anything, so you're allocating to get the chars/string out, which you need to do for hashcode or equals. Newer JVMs don't use StringBuilder at all for string concat, though there's a CLI option to enable it.

Names are hard and all the good names are taken! Short names are nice, but less descriptive.

tommyettinger reviewed (dismissed):
Looks good; I like the more consistent logging (without println() mixed in). The new name is also better for the IntMap. Approved!

JohnLavender474 commented:
@NathanSweet Whenever you have time, would you mind looking over this PR to see if it's good?

tommyettinger commented:
Looks like you need to run spotlessApply again; something changed in formatting somewhere and spotless doesn't like it.

NathanSweet commented:
Tiled is one of the areas I'm unfamiliar with, having never looked in there nor implemented an app using it. I can do a once over though! I'll dump some opinions or just notes, feel free to disagree, no worries.

* `objectIdToPropsMap` I'd drop the type in the name -> `objectIdToProps`. If `objectId` is not more descriptive in a Tiled context I'd use `idToProps`.
* I guess `runOnEndOfLoadTiledMap` and `objectIdToPropsMap` can be null at the end of `loadTiledMap`.
* Diamond operator on RHS doesn't do anything.
* Multiline comments sure take up a lot of lines when the start and end of the comment get their very own line. I'd just use `//` and let the IDE wrap it at 130 columns.
* @Override on JRE classes (Runnable) isn't useful, they are never going to make a breaking change.
* Queue is cool but you `addFirst` and `removeFirst` (LIFO), would be the same with Array `add` and `pop`. Could just iterate and null `runOnEndOfLoadTiledMap` afterward.
* The runnables seem OK. Another way could be to collect properties, name, id triplets but it'd be equivalent.
* Why catch and log? Why not catch and throw? If I need object properties and they fail, my app should probably die. Reacting to the log message is difficult.
* Log or exception could include the invalid value.
* MapObject vs MapProperties, I'm afraid I'm no help there.
* Storing ID to properties for everything may be a lot of storing. That's a little extra work and memory during loading only, seems pretty fine. I'm guessing there's no existing "get by ID" mechanism this could use?

Overall it seems good. The nulling and logging items may warrant revision.

tommyettinger commented:
> * Queue is cool but you `addFirst` and `removeFirst` (LIFO), would be the same with Array `add` and `pop`. Could just iterate and null `runOnEndOfLoadTiledMap` afterward.

`Array.add()` is the same as `addLast()`, no? Adding at the start of an `Array` means every existing element needs to get shifted over, but for a `Queue` it just changes where the head is.

The implementation for `Queue` is nice and I learned a lot reading it. I also figured out how to make a similar class (also a double-ended queue) have O(1) random access; I can't remember if `Queue` had this or had part of this, but [my ObjectDeque class](https://github.com/tommyettinger/jdkgdxds/blob/master/src/main/java/com/github/tommyettinger/ds/ObjectDeque.java#L1184-L1201) allows getting and setting elements by index in constant time. Removing and inserting in the middle are O(n), like `Array`.

JohnLavender474 commented:
@NathanSweet @tommyettinger After thinking about it for a while, I decided that instead of fetching MapProperties, the "object" type should return MapObject instead. It make more sense that users would want to have direct access to the MapObject being pointed to. I changed the code in the BaseTmxMapLoader and included a public method for users to fetch a copy of the id-to-object map. And the test class has been changed too; it includes better logging and repeats the test 3 times to make sure that reusing the same TmxMapLoader instance doesn't present any problems. I edited the Tiled map to include two object layers to ensure that objects referencing each other across layers doesn't cause issues.

@NathanSweet I made most of the changes you suggested. Just a few minor quibbles: I consider it good practice to include the diamond operator and the @Override tag.

SimonIT commented on gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java:
```diff
@@ -88,6 +86,16 @@ public Array<AssetDescriptor> getDependencies (String fileName, FileHandle tmxFi
+	 *
+	 * @return the map of the object ids to the {@link MapObject} instances, or null if the load tiled map method has not been
+	 *         called yet. */
+	public IntMap<MapObject> getIdToObject () {
```
We have a `@Null` annotation for nullable return values

JohnLavender474 replied:
Added the `@Null tag`

NathanSweet commented on gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java:
```diff
@@ -83,6 +85,16 @@ public Array<AssetDescriptor> getDependencies (String fileName, FileHandle tmxFi
+	 * @return the map of the object ids to the {@link MapObject} instances, or null if the load tiled map method has not been
+	 *         called yet. */
+	public @Null IntMap<MapObject> getIdToObject () {
+		return idToObject == null ? null : new IntMap<>(idToObject);
```
Why a copy?

JohnLavender474 replied:
It returns a copy so that the internal state of the map can't be changed from the outside. The user should call this method only once each time the loadTiledMap method is called so that it doesn't incur much runtime penalty.

JohnLavender474 replied:
I don't have a strong opinion about this though if you think it should just return the internal map instead of a copy

JohnLavender474 replied:
I think you make a great argument for why it shouldn't return a copy. My thought was to protect the internal state, but after reading your explanation, I agree there's not much point since the map is reallocated every time the load tiled map method is called. I've changed it.

NathanSweet commented on gdx/src/com/badlogic/gdx/maps/tiled/BaseTmxMapLoader.java:
```diff
@@ -431,22 +450,44 @@ protected void loadObject (TiledMap map, MapObjects objects, Element element, fl
+						Runnable fetch = new Runnable() {
+							@Override
+							public void run () {
+								MapObject props = idToObject.get(id);
```
props -> object

JohnLavender474 replied:
Good catch

NathanSweet reviewed (dismissed):
Just those comments, looks good to me.

NathanSweet commented:
> Array.add() is the same as addLast(), no? Adding at the start of an Array means every existing element needs to get shifted over, but for a Queue it just changes where the head is.

@tommyettinger Right, but this usage was Queue `addFirst` and `removeFirst`, so last-in-first-out. Array `add` and `pop` is also last-in-first-out, so functionally the same (and maybe the teeniest bit more familiar to most devs).

tommyettinger reviewed (dismissed):
I don't see anything wrong here, myself. The CHANGES file needs to have its conflict resolved, which should be possible to do in the web editor or your Git client of choice. I think we can get this merged before the next release (or very soon afterward if the release happens sooner than I'm thinking it will); this PR has been checked over lots already.

JohnLavender474 commented:
> I don't see anything wrong here, myself. The CHANGES file needs to have its conflict resolved, which should be possible to do in the web editor or your Git client of choice. I think we can get this merged before the next release (or very soon afterward if the release happens sooner than I'm thinking it will); this PR has been checked over lots already.

@tommyettinger  CHANGES file should be resolved now. Awesome to here it might get merged! I'd like to make contributions to open-source projects something I do on a regular basis.

tommyettinger reviewed (approved):
Just reiterating my approval now that there are no conflicts.
